packages/dds/tree/src/forest/cursor.ts

```diff
@@ -4,89 +4,183 @@
+     *
+     * NOT allowed if mode is `Fields` and also `pending`.
+     */
+    seek(offset: number): boolean;
 
-/**
- * A stateful low-level interface for reading tree data.
- *
- * TODO: Needs rules around invalidation/mutation of the underlying tree.
- * Should either be documented here, or each producer should document them
- * (and likely via returning a sub-interface with documentation on the subject).
- *
- * TODO: Needs a way to efficiently clone Cursor so patterns like lazy tree reification can be implemented efficiently.
- *
- * TODO: add optional fast path APIs for more efficient handling when supported by underlying format and reader.
- * Leverage "chunks" and "shape" for this, and skip to next chunk with seek (chunk length).
- * Default chunks of size 1, and "node" shape?
- */
-export interface ITreeCursor<TResult = TreeNavigationResult> {
-    /** Select the child located at the given key and index. */
-    down(key: FieldKey, index: number): TResult;
+    // ********** APIs for when mode = Fields, and not pending ********** //
 
     /**
-     * Moves `offset` entries in the field.
+     * Returns the FieldKey for the current field.
+     *
+     * Allowed when `mode` is `Fields`, and not `pending`.
      */
-    seek(offset: number): TResult;
+    getCurrentFieldKey(): FieldKey;
 
-    /** Select the parent of the currently selected node. */
-    up(): TResult;
+    /**
+     * @returns the number of immediate children in the current field.
+     *
+     * Allowed when `mode` is `Fields`, and not `pending`.
+     */
+    getCurrentFieldLength(): number;
 
-    /** The type of the currently selected node. */
-    readonly type: TreeType;
+    /**
+     * Sets current node to the on at the provided `index` of the current field.
+     *
+     * Allowed when `mode` is `Fields`, and not `pending`.
+     * Sets mode to `Nodes`.
+     */
+    enterChildNode(index: number): void;
+
+    // ********** APIs for when mode = Nodes ********** //
 
     /**
-     * @returns the keys of the currently selected node.
-     * TODO: ordering invariants: Consistent over time? Consistent across nodes? Sorted?
-     * TODO: empty fields: are they always omitted here? Sometimes omitted? Depends on field kind and schema?
-     * */
-    keys: Iterable<FieldKey>;
+     * @returns a path to the current node.
+     *
+     * Only valid when `mode` is `Nodes`.
+     */
+    getPath(): UpPath;
 
-    /** @returns the number of immediate children for the given key of the currently selected node. */
-    length(key: FieldKey): number;
+    /**
+     * Index (with its parent field) of the current node.
+     *
+     * Only valid when `mode` is `Nodes`.
+     */
+    readonly currentIndexInField: number;
 
-    /** value associated with the currently selected node. */
+    /**
+     * Index (with its parent field) of the first node in the current chunk.
+     * Always less than or equal to `currentIndexInField`.
```

Comment:
```suggestion
     * Always greater than or equal to `currentIndexInField`.
```
Maybe I'm not understanding chunking right, but this is what I would have expected.

Reply:
This was right before. A field's contents are made of chunks, and the current node is in one of those chunks. The chunk the current node is in starts at or before the current node, not after it.

Reply:
Thanks, this makes sense. I got confused.